ipam/ipam.go

```diff
@@ -0,0 +1,128 @@
+package ipam
+
+import (
+	"context"
+
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	"github.com/giantswarm/e2etests/ipam/provider"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+)
+
+type Config struct {
+	GuestFramework *framework.Guest
+	Logger         micrologger.Logger
+	Provider       provider.Interface
+}
+
+type IPAM struct {
+	guestFramework *framework.Guest
+	logger         micrologger.Logger
+	provider       provider.Interface
+}
+
+func New(config Config) (*IPAM, error) {
+	if config.GuestFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GuestFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+
+	s := &IPAM{
+		guestFramework: config.GuestFramework,
+		logger:         config.Logger,
+		provider:       config.Provider,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+	var err error
+
+	const (
+		clusterOne   = "cluster0"
+		clusterTwo   = "cluster1"
+		clusterThree = "cluster2"
+		clusterFour  = "cluster3"
+	)
+
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	// First create all three guest clusters
+	for _, cn := range clusters {
+		err = c.provider.RequestGuestClusterCreation(cn)
```

Comment:
Why not just `CreateCluster`? I also would like to have logging before and after the action so this would be more self explanatory in consequence.

Reply:
`Request...` because I wanted to emphasize the fact that this function will not wait for creation but only request for a new cluster (which creation of provider specific config kind of is). I don't have strong opinion here though, if you think `CreateCluster` is good despite the fact it's not waiting for cluster creation.

I'll add logging.

Reply:
Lets rather have godoc on the provider interface and if necessary in the code here to make certain things clear. The function name cannot do this job I think.